chainerx_cc/chainerx/routines/misc.h

```diff
@@ -25,4 +25,9 @@ Array Minimum(const Array& x1, Scalar x2);
 Array Minimum(Scalar x1, const Array& x2);
 Array Minimum(const Array& x1, const Array& x2);
 
+Array Clip(const Array& x1, Scalar min, Scalar max);
+Array Clip(const Array& x1, const Array& a_min, Scalar max);
+Array Clip(const Array& x1, Scalar min, const Array& a_max);
+Array Clip(const Array& x1, const Array& min, const Array& max);
```

Comment:
We're usually okay with propagating optional arguments to the routines (C++ interface). Do you have a strong reason not to do so in this case, basically (with acceptable overhead) moving the error checking from the Python glue code to C++. The benefit is that the glue code becomes simpler and the interfaces more similar.

Reply:
If the glue code requires more than 10 line I would use optional typed argument but it's quite short.
Polluting overloads of `Clip` function seems more complex to me. If I'm using ChainerX directly from C++ I don't want many unused overloads around the error messages.
In this case keeping complexity to glue code seems to make the total complexity less to me.

Though if the glue code to the python is generate fully automatic I'll define the optional interface.
And it's not too late to define numpy like interface after a demand of somebody.

Reply:
Thanks for your explanation. I see your point and also think it's important to think of the C++ and Python interfaces separately. It does make sense to keep the interface non-optional for the C++ especially in terms of the code quality of the implementation but I'm at the same time not sure if the average (Python +)C++ user is happier having to write the current glue code logic themselves.